### PR TITLE
CMP-1108: Deprecate resources in `compose.ui` in favour of the new resource library

### PR DESCRIPTION
**Pull request by terrakok**
To work with resources in the desktop compose the recommended way is to use the special resources library.

Fixes https://youtrack.jetbrains.com/issue/CMP-1108

### Migration snippet
To read java resources by path as Painters you may use:
```kotlin
@Composable
internal fun painterResource(
    resourcePath: String
): Painter = when (resourcePath.substringAfterLast(".")) {
    "svg" -> rememberSvgResource(resourcePath)
    "xml" -> rememberVectorXmlResource(resourcePath)
    else -> rememberBitmapResource(resourcePath)
}

@Composable
internal fun rememberBitmapResource(path: String): Painter {
    return remember(path) { BitmapPainter(readResourceBytes(path).decodeToImageBitmap()) }
}

@Composable
internal fun rememberVectorXmlResource(path: String): Painter {
    val density = LocalDensity.current
    val imageVector = remember(density, path) { readResourceBytes(path).decodeToImageVector(density) }
    return rememberVectorPainter(imageVector)
}

@Composable
internal fun rememberSvgResource(path: String): Painter {
    val density = LocalDensity.current
    return remember(density, path) { readResourceBytes(path).decodeToSvgPainter(density) }
}

private object ResourceLoader
private fun readResourceBytes(resourcePath: String) =
    ResourceLoader.javaClass.classLoader.getResourceAsStream(resourcePath).readAllBytes()
```

### Breaking changes - Resources
- Deprecate resources in `compose.ui` in favour of the new resource library

### CODE REVIEW

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/res/DesktopSvgResources.desktop.kt
Comment (igordmn) [inline compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/res/DesktopSvgResources.desktop.kt:50]: > Chore - Resources

Let's change it to
```
Breaking changes - Resources
```

The values should be from this [list](https://github.com/JetBrains/compose-multiplatform-core/edit/jb-main/.github/PULL_REQUEST_TEMPLATE.md) (we extend it when needed)

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/res/DesktopSvgResources.desktop.kt
Comment (igordmn) [inline compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/res/DesktopSvgResources.desktop.kt]:
```suggestion
@Deprecated("Migrate to the Compose resources library. See https://www.jetbrains.com/help/kotlin-multiplatform-dev/compose-images-resources.html")
```

Comment (igordmn) [inline compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/res/DesktopSvgResources.desktop.kt]: Let's check how resources work in [InteliJ plugin integration](https://github.com/JetBrains/compose-multiplatform/tree/master/examples/intellij-plugin) before deprecation.

Comment (igordmn) [inline compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/res/DesktopSvgResources.desktop.kt]: Also, we need to check if the new resources can be used in our current/potential client environments (Toolbox, InteliJ, Android Studio repos). Not sure that they apply `compose-multiplatform` Gradle plugin. Could you investigate this?

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/res/PainterResources.desktop.kt
Comment (MatkovIvan) [inline compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/res/PainterResources.desktop.kt]: We don't need to 
1. Remove experimental
2. Deprecate experimental (just remove)

**Comment by terrakok** [comment]
Checked with Compose 1.7.0 and Kotlin 2.0.20.
Found two problems: 
 - [CMP-6543](https://youtrack.jetbrains.com/issue/CMP-6543) java.lang.NoSuchMethodError: 'void java.io.InputStream.skipNBytes(long)' **Fixed**
 - [CMP-6541](https://youtrack.jetbrains.com/issue/CMP-6541) Compose resource accessors aren't generated when Kotlin JVM plugin is applied **there is easy workaround**

image

(button's texts, scroll items and a logo use compose resources)

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/res/DesktopSvgResources.desktop.kt
Comment (terrakok) [reply compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/res/DesktopSvgResources.desktop.kt]: Since compose 1.7.0 there are functions to convert xml, svg, png etc. byte arrays:
```kotlin
fun ByteArray.decodeToImageBitmap(): ImageBitmap
fun ByteArray.decodeToImageVector(density: Density): ImageVector
fun ByteArray.decodeToSvgPainter(density: Density): Painter
```

if an app doesn't use gradle plugin it is still able to read a plain java resource as the byte array and convert it to the compose object

Comment (terrakok) [reply compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/res/DesktopSvgResources.desktop.kt]: commented below

Comment (terrakok) [reply compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/res/DesktopSvgResources.desktop.kt]: ✅

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/res/DesktopSvgResources.desktop.kt
Comment (terrakok) [reply compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/res/DesktopSvgResources.desktop.kt:50]: ✅

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/res/PainterResources.desktop.kt
Comment (terrakok) [reply compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/res/PainterResources.desktop.kt]: We cannot delete these function because they are being used in non experimental public API 🤷 
Example:
```kotlin
//non experimental
fun painterResource(resourcePath: String, loader: ResourceLoader): Painter

//non experimental
private fun rememberSvgResource(
    resourcePath: String, 
    loader: ResourceLoader = ResourceLoader.Default
): Painter

//experimental
internal inline fun <T> useResource(
    resourcePath: String,
    loader: ResourceLoader,
    block: (InputStream) -> T
): T
```

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/res/DesktopSvgResources.desktop.kt
Comment (igordmn) [reply compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/res/DesktopSvgResources.desktop.kt]: Let's write a replacement function in the deprecation comment, to be completely sure, and to provide a fast solution for non-Gradle users

**Review by MatkovIvan — COMMENTED** [review]
Removing experimental marks still doesn't LGTM. We can just remove it instead of committing for endless "deprecation" compatibility support

**Comment by igordmn** [comment]
> Removing experimental marks still doesn't LGTM. We can just remove it instead of committing for endless "deprecation"

We can do that for the experimental API, but we should move the implementation inside non-experimental API

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/res/DesktopSvgResources.desktop.kt
Comment (igordmn) [reply compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/res/DesktopSvgResources.desktop.kt]: @terrakok, I am waiting for this to approve the PR.

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/res/PainterResources.desktop.kt
Comment (igordmn) [inline compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/res/PainterResources.desktop.kt:55]: Besides `painterResource`, we need to deprecate in another PR:
```
fun Font(
    resource: String,
    weight: FontWeight = FontWeight.Normal,
    style: FontStyle = FontStyle.Normal
): Font = ResourceFont(resource, weight, style)
```

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/res/DesktopSvgResources.desktop.kt
Comment (igordmn) [reply compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/res/DesktopSvgResources.desktop.kt]: We also need a replacement for `painterResource`, if users want to load specifically from classpath (Jewel, IntelliJ platform, Swing applications).

I suggest do it this way:

1. Add "If you need to load resources specificly from Java classpath, you should still use the new resource library and use a snippet from https://github.com/JetBrains/compose-multiplatform-core/pull/1457"
2. Add a snippet in the PR description. Something like this:
```
@Composable
fun bitmapPainterResource(path: String): Painter =
    remember { BitmapPainter(readResourceBytes(path).decodeToImageBitmap()) }

@Composable
fun imageVectorPainterResource(path: String): Painter {
    val imageVector = remember { readResourceBytes(path).decodeToImageVector() }
    return rememberVectorPainter(imageVector)
}

@Composable
fun svgPainterResource(path: String): Painter =
    remember { readResourceBytes(path).decodeToSvgPainter() }

fun readResourceBytes(resourcePath: String) =
    ClassLoader.getSystemClassLoader().getResourceAsStream(resourcePath).readAllBytes()
```
(please, check correctness)

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/res/PainterResources.desktop.kt
Comment (MatkovIvan) [reply compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/res/PainterResources.desktop.kt]: Can we make it internal then?

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/res/Resources.desktop.kt
Comment (igordmn) [inline compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/res/Resources.desktop.kt]: It is internal. Can we just remove Deprecated and Experimental?

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/res/Resources.desktop.kt
Comment (igordmn) [inline compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/res/Resources.desktop.kt]: We also can mark it internal, and the function that uses it

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/res/PainterResources.desktop.kt
Comment (igordmn) [reply compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/res/PainterResources.desktop.kt:55]: @terrakok, we need to do that as soon as possible. `painterResource` and `Font` technically of one kind of features, better to deprecated them in one release.

Comment (terrakok) [reply compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/res/PainterResources.desktop.kt:55]: Why? The `resource: String` is not a file path here. It is a name in the system

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/res/Resources.desktop.kt
Comment (igordmn) [inline compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/res/Resources.desktop.kt]:
```suggestion
```

**Review by igordmn — APPROVED** [review]

**Comment by igordmn** [comment]
Let's wait for a second approval

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/res/PainterResources.desktop.kt
Comment (igordmn) [reply compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/res/PainterResources.desktop.kt:55]: It is actually a file path. See https://github.com/JetBrains/compose-multiplatform-core/blob/eece65f88187f5cb234144ce84cb87e38260a1c3/compose/desktop/desktop/samples/src/jvmMain/kotlin/androidx/compose/desktop/examples/example1/Main.jvm.kt#L152

**Review by MatkovIvan — APPROVED** [review]

**Comment by Nohus** [comment]
@terrakok I believe this broke `org.jetbrains.compose.animatedimage.loadResourceAnimatedImage`. You can no longer load gifs from `composeResources/files/image.gif`. You have to instead move them to the normal non-Compose `resources/image.gif`. That's an unannounced breaking change in `org.jetbrains.compose.components:components-animatedimage-desktop:1.7.0-beta02` that I found out about by user reports of crashing.

**Comment by terrakok** [comment]
I checked it with the code and it works:
```kotlin
@Composable
internal fun App() = AppTheme {
    val (gif, setGif) = remember { mutableStateOf<AnimatedImage?>(null) }
    LaunchedEffect(Unit) {
        setGif(loadAnimatedImage(Res.getUri("files/olaf.gif")))
    }
    if (gif != null) {
        Image(
            modifier = Modifier.size(100.dp),
            bitmap = gif.animate(),
            contentDescription = null
        )
    }
}
```

If you have a problem then file an issue on the YouTrack, please

**Pull request merged**
